libbpf-tools/sigsnoop.c

```diff
@@ -0,0 +1,266 @@
+	}
+	return 0;
+}
+
+static void alias_parse(char *prog)
+{
+	char *name = basename(prog);
+
+	if (!strcmp(name, "killsnoop")) {
+		kill_only = true;
+	}
+}
+
+static void sig_int(int signo)
+{
+	exiting = 1;
+}
+
+static void handle_event(void *ctx, int cpu, void *data, __u32 data_sz)
+{
+	struct event *e = data;
+	time_t t;
+	struct tm *tm;
+	char ts[32];
+
+	time(&t);
+	tm = localtime(&t);
+	strftime(ts, sizeof(ts), "%H:%M:%S", tm);
+	if (signal_name)
+		printf("%-8s %-7d %-16s %-9s %-7d %-6d\n",
+		       ts, e->pid, e->comm, sig_name[e->sig], e->tpid, e->ret);
+	else
+		printf("%-8s %-7d %-16s %-9d %-7d %-6d\n",
+		       ts, e->pid, e->comm, e->sig, e->tpid, e->ret);
+}
+
+static void handle_lost_events(void *ctx, int cpu, __u64 lost_cnt)
+{
+	warn("lost %llu events on CPU #%d\n", lost_cnt, cpu);
+}
+
+int main(int argc, char **argv)
+{
+	static const struct argp argp = {
+		.options = opts,
+		.parser = parse_arg,
+		.doc = argp_program_doc,
+	};
+	struct perf_buffer_opts pb_opts;
+	struct perf_buffer *pb = NULL;
+	struct sigsnoop_bpf *obj;
+	int err;
+
+	alias_parse(argv[0]);
+	err = argp_parse(&argp, argc, argv, 0, NULL, NULL);
+	if (err)
+		return err;
+
+	err = bump_memlock_rlimit();
+	if (err) {
+		warn("failed to increase rlimit: %d\n", err);
+		return 1;
+	}
+
+	obj = sigsnoop_bpf__open();
+	if (!obj) {
+		warn("failed to open BPF object\n");
+		return 1;
+	}
+
+	obj->rodata->target_pid = target_pid;
+	obj->rodata->target_signal = target_signal;
+	obj->rodata->failed_only = failed_only;
+
+	if (kill_only) {
+		bpf_program__set_autoload(obj->progs.sig_trace, false);
+	} else {
+		bpf_program__set_autoload(obj->progs.kill_entry, false);
+		bpf_program__set_autoload(obj->progs.kill_exit, false);
+		bpf_program__set_autoload(obj->progs.tkill_entry, false);
+		bpf_program__set_autoload(obj->progs.tkill_exit, false);
+		bpf_program__set_autoload(obj->progs.tgkill_entry, false);
+		bpf_program__set_autoload(obj->progs.tgkill_exit, false);
+	}
+
+	err = sigsnoop_bpf__load(obj);
+	if (err) {
+		warn("failed to load BPF object: %d\n", err);
+		goto cleanup;
+	}
+
+	err = sigsnoop_bpf__attach(obj);
+	if (err) {
+		warn("failed to attach BPF programs: %d\n", err);
+		goto cleanup;
+	}
+
+	pb_opts.sample_cb = handle_event;
+	pb_opts.lost_cb = handle_lost_events;
+	pb = perf_buffer__new(bpf_map__fd(obj->maps.events), PERF_BUFFER_PAGES, &pb_opts);
+	err = libbpf_get_error(pb);
+	if (err) {
+		warn("failed to open perf buffer: %d\n", err);
+		goto cleanup;
+	}
+
+	if (signal(SIGINT, sig_int) == SIG_ERR) {
+		warn("can't set signal handler: %s\n", strerror(-errno));
+		goto cleanup;
+	}
+
+	printf("%-8s %-7s %-16s %-9s %-7s %-6s\n",
+	       "TIME", "PID", "COMM", "SIG", "TPID", "RESULT");
+
+	while (1) {
+		err = perf_buffer__poll(pb, PERF_POLL_TIMEOUT_MS);
+		if (err == -EINTR) {
+ 			err = 0;
+ 			goto cleanup;
+ 		}
+
+		if (err < 0)
+			break;
+		if (exiting)
+			goto cleanup;
+	}
+	warn("error polling perf buffer: %d\n", err);
```

Comment:
It's again me :grimacing:, could you please follow the same approach as in https://github.com/iovisor/bcc/pull/3630?